pkg/storer/internal/reserve/reserve.go

```diff
@@ -122,6 +122,93 @@ func (r *Reserve) Put(ctx context.Context, chunk swarm.Chunk) error {
 		if err != nil {
 			return fmt.Errorf("load or store stamp index for chunk %v has fail: %w", chunk, err)
 		}
+
+		sameAddressOldChunkStamp, err := chunkstamp.LoadWithBatchID(s.IndexStore(), reserveNamespace, chunk.Address(), chunk.Stamp().BatchID())
+		if err != nil && !errors.Is(err, storage.ErrNotFound) {
+			return err
+		}
+
+		// same address
+		if sameAddressOldChunkStamp != nil {
+			sameAddressOldStampIndex, err := stampindex.LoadWithStamp(s.IndexStore(), reserveNamespace, sameAddressOldChunkStamp)
```

Comment:
`sameAddressOldStampIndex` was loaded so that we can read the `hash` immediately instead of hashing the stamp again.

Reply:
hm, I've a the feeling the hashing would be faster than a database read.

Reply:
and is it not the same as `oldItem`?

Reply:
`sameAddressOldChunkStamp` does not ensure that the chunk retrieved has also the same stamp (not only batchId), or does that? (I don't see clear thing regarding that in the `LoadWithBatchID`'s query)

Reply:
> hm, I've a the feeling the hashing would be faster than a database read.

Maybe. However, reading the stored hash is better than computing it again because reading is resistant to changes in the hash function.

> sameAddressOldChunkStamp does not ensure that the chunk retrieved has also the same stamp (not only batchId), or does that? (I don't see clear thing regarding that in the LoadWithBatchID's query)

I'm not sure what you mean but `LoadWithBatchID` just finds the `stamp` associated with the chunk address and batchID

Reply:
> However, reading the stored hash is better than computing it again because reading is resistant to changes in the hash function.

If something changes in the hash function all db records regarding that should be updated anyway, shouldn't it?

> I'm not sure what you mean but `LoadWithBatchID` just finds the `stamp` associated with the chunk address and batchID

I mean that the chunk which stored and you want to override is retrieved by only the chunk address and batchID but this chunk should be replaced only if their stamp indices are the same as well (since if they are on different index they can be seem as legit payed stores).
My argue here is `sameAddressOldChunkStamp` returns a stamp which may not satisfy the stamp index match with the one that the inflight chunk has <-> it returns a stored chunk with a legit stamp and it wants to replace with the new chunk with a different/unrelated legit stamp.

It does not cover the situation when one uploads different payloads with the same batchID on the same SOC address used up multiple valid slots.

Before the same address condition there should be a search for retrieving all chunks with same chunk address and batch id and continue the replacement logic if stamp indices match.

Reply:
> It does not cover the situation when one uploads different payloads with the same batchID on the same SOC address used up multiple valid slots.

We don't store multiple payloads for the same chunk address. This PR is to override the old one with the new one.

Reply:
but then why is the same batchId condition at chunk retrieval on line 126?